### PR TITLE
Vision: Add gRPC support for safe search.

### PR DESCRIPTION
**Pull request by daspecster**
~~Based from #2937~~

There's a few things I don't like about this that I'll comment on. I'm not sure how to solve them right now. Maybe you guys have some ideas?

### CODE REVIEW

File: vision/unit_tests/test_annotations.py [hunk 1]
```diff
@@ -67,17 +67,26 @@ def test_ctor(self):
...
         annotations = self._make_one().from_pb(image_response)
         self.assertEqual(annotations.labels, [])
         self.assertEqual(annotations.logos, [])
-        self.assertEqual(annotations.faces, ())
+        self.assertEqual(annotations.faces, [])
         self.assertEqual(annotations.landmarks, [])
         self.assertEqual(annotations.texts, [])
-        self.assertEqual(annotations.safe_searches, ())
-        self.assertEqual(annotations.properties, ())
+        self.assertEqual(len(annotations.safe_searches), 1)
```
Comment (daspecster) [inline vision/unit_tests/test_annotations.py]: On the other protobuf objects, there has been a way for me to tell if there are results or not. In this case, the values for `adult`, `spoof` etc... are defaulting to`0` which is the valid `Likelihood.UNKNOWN` type.

I would have to track the feature type's that are passed in from the top in order to determine if this was a requested feature type or not. There has to be a better way.

**Comment by daspecster** [comment]
@dhermes this should be ready to go. Let me know if you have any feedback.

File: vision/google/cloud/vision/safe.py [hunk 2]
```diff
@@ -54,14 +54,30 @@ def from_api_repr(cls, response):
...
+        :type image: :class:`~google.cloud.grpc.vision.v1.image_annotator_pb2.\
+                      SafeSearchAnnotation`
+        :param image: Protobuf response from Vision API with safe search data.
+
+        :rtype: :class:`~google.cloud.vision.safe.SafeSearchAnnotation`
+        :returns: Instance of ``SafeSearchAnnotation``.
+        """
+        classifications = map(_get_pb_likelihood, [image.adult, image.spoof,
+                                                   image.medical,
+                                                   image.violence])
```
Comment (dhermes) [inline vision/google/cloud/vision/safe.py]: Style nit: Suggest instead

```python
values = [image.adult, image.spoof, image.medical, image.violence]
classifications = map(_get_pb_likelihood, values)
```

File: vision/unit_tests/test_annotations.py [hunk 3]
```diff
@@ -76,9 +78,16 @@ def test_from_pb(self):
         self.assertEqual(annotations.faces, [])
         self.assertEqual(annotations.landmarks, [])
         self.assertEqual(annotations.texts, [])
-        self.assertEqual(annotations.safe_searches, ())
         self.assertIsNone(annotations.properties)
 
+        self.assertIsInstance(annotations.safe_searches, SafeSearchAnnotation)
+        safe_search = annotations.safe_searches
+        unknown = Likelihood.UNKNOWN
+        self.assertEqual(safe_search.adult, unknown)
```
Comment (dhermes) [inline vision/unit_tests/test_annotations.py]: Wouldn't `assertIs` be better for an enum?

File: vision/unit_tests/test_safe.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
+# Copyright 2016 Google Inc.
```
Comment (dhermes) [inline vision/unit_tests/test_safe.py]: High-level

1. If the `safe` module already existed, this test should have also already existed. Transitive test coverage is not the same as writing actual tests
1. If `safe` was small enough to get transitive coverage without every directly testing anything in it, you probably didn't need to give it it's on module
1. :-1: to the module name `safe`, it is unclear

File: vision/unit_tests/test_safe.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+class TestSafeSearchAnnotation(unittest.TestCase):
+    @staticmethod
+    def _get_target_class():
+        from google.cloud.vision.safe import SafeSearchAnnotation
+        return SafeSearchAnnotation
+
+    def test_safe_search_annotation(self):
+        from google.cloud.vision.likelihood import Likelihood
+        from unit_tests._fixtures import SAFE_SEARCH_DETECTION_RESPONSE
```
Comment (dhermes) [inline vision/unit_tests/test_safe.py]: Do you really need `_fixtures`?

File: docs/vision-usage.rst [hunk 6]
```diff
@@ -224,8 +224,7 @@ categorize the entire contents of the image under four categories.
     >>> client = vision.Client()
     >>> with open('./image.jpg', 'rb') as image_file:
     ...     image = client.image(content=image_file.read())
-    >>> safe_search_results = image.detect_safe_search()
-    >>> safe_search = safe_search_results[0]
+    >>> safe_search = image.detect_safe_search()
```
Comment (dhermes) [inline docs/vision-usage.rst]: API didn't change, these docs were just stale?

File: vision/unit_tests/test_safe.py (hunk 5 above)
Comment (daspecster) [inline vision/unit_tests/test_safe.py]: For safe search, it could probably be a local variable. However I like having it all in the same place. Some of the other responses are fairly large. I guess it just depends if you want the tests or not. If so, then we may have to bump the max module lines value again.

File: vision/unit_tests/test_safe.py (hunk 4 above)
Comment (daspecster) [inline vision/unit_tests/test_safe.py]: I can rename to `safe_search`.

Comment (dhermes) [inline vision/unit_tests/test_safe.py]: Yes please do

File: vision/unit_tests/test_safe.py (hunk 5 above)
Comment (dhermes) [inline vision/unit_tests/test_safe.py]: Bumping that value doesn't bother me.

> I like having it all in the same place

It may just be a preference difference between us? ISTM that it can be said that individual tests are hard to reason about if the data is not in the test (independent of preference), but maybe I'm wrong?

File: docs/vision-usage.rst (hunk 6 above)
Comment (daspecster) [inline docs/vision-usage.rst]: Yup, stale.

File: vision/unit_tests/test_safe.py (hunk 5 above)
Comment (daspecster) [inline vision/unit_tests/test_safe.py]: Sure, I agree with that. However in this case I don't really like having [this](https://github.com/GoogleCloudPlatform/google-cloud-python/blob/422378126b465ce26795fd159eed625f86d996a0/vision/unit_tests/_fixtures.py#L278-L1688) much in a test case.

I guess I could test it in parts instead of a full response. Then we would be relying more on the system test to make sure that it works in its entirety.

Comment (daspecster) [inline vision/unit_tests/test_safe.py]: @dhermes did you want me to bring the safe search response into the tests in this PR or do them all together in another PR?

Comment (dhermes) [inline vision/unit_tests/test_safe.py]: Whoa that's huge. Don't worry about that change

File: vision/unit_tests/test_safe_search.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
+# Copyright 2016 Google Inc.
```
Comment (dhermes) [inline vision/unit_tests/test_safe_search.py]: s/2016/2017

**Comment by daspecster** [comment]
Updated copyright and squashed.
